### PR TITLE
*: Refactor aggregation

### PR DESCRIPTION
**Pull request by wangyisong1996**
Aggregation results were stored in aggregation functions, so one map is created for each function to store different groups’ results, which requires much memory.
The aggregation functions are refactored and are now stateless, and the maps are kept in executor (`executor.HashAggExec` and `mocktikv.aggregateExec`) so that only one map is needed to store all the functions’ results, which reduced the memory usage.

### CODE REVIEW

**Comment by wangyisong1996** [comment]
@hanfei1991 
@zz-jason 
PTAL

File: expression/aggregation/aggregation.go [hunk 1]
```diff
@@ -256,62 +237,16 @@ func (af *aggFunction) SetArgs(args []expression.Expression) {
...
-		return errors.Trace(err)
-	}
-	ctx.Count++
-	return nil
+	return ctx
 }
 
-func (af *aggFunction) streamUpdateSum(row []types.Datum, sc *variable.StatementContext) error {
-	ctx := af.getStreamedContext()
+func (af *aggFunction) updateSum(row []types.Datum, ctx *AggEvaluateContext, sc *variable.StatementContext) error {
```
Comment (zz-jason) [inline expression/aggregation/aggregation.go]: ditto

File: executor/aggregate.go [hunk 2]
```diff
@@ -135,12 +138,25 @@ func (e *HashAggExec) innerNext() (ret bool, err error) {
...
-		af.Update(srcRow, groupKey, e.sc)
+	aggCtxs := e.getContexts(groupKey)
+	for i, af := range e.AggFuncs {
+		af.Update(srcRow, aggCtxs[i], e.sc)
 	}
 	return true, nil
 }
 
+func (e *HashAggExec) getContexts(groupKey []byte) []*aggregation.AggEvaluateContext {
+	aggCtxs, ok := e.aggCtxsMap[string(groupKey)]
```
Comment (zz-jason) [inline executor/aggregate.go]: extract `string(groupKey)` to a variable

File: expression/aggregation/avg.go [hunk 3]
```diff
@@ -102,24 +95,12 @@ func (af *avgFunction) calculateResult(ctx *aggEvaluateContext) (d types.Datum)
 	return
 }
 
-// GetGroupResult implements Aggregation interface.
-func (af *avgFunction) GetGroupResult(groupKey []byte) types.Datum {
-	ctx := af.getContext(groupKey)
+// GetResult implements Aggregation interface.
+func (af *avgFunction) GetResult(ctx *AggEvaluateContext) types.Datum {
```
Comment (zz-jason) [inline expression/aggregation/avg.go]: we can remove function `calculateResult()` and move the calculate logic to function `GetResult()`

File: expression/aggregation/aggregation.go [hunk 4]
```diff
@@ -32,27 +32,23 @@ import (
 type Aggregation interface {
 	fmt.Stringer
 	json.Marshaler
+	// // Update during executing.
+	// Update(row []types.Datum, groupKey []byte, sc *variable.StatementContext) error
 	// Update during executing.
-	Update(row []types.Datum, groupKey []byte, sc *variable.StatementContext) error
+	Update(row []types.Datum, ctx *AggEvaluateContext, sc *variable.StatementContext) error
```
[the remark this answers is not in the document]
Comment (wangyisong1996) [reply expression/aggregation/aggregation.go]: The original `Update` function is `Update(row []types.Datum, groupKey []byte, sc *variable.StatementContext)` and I'm just changing the `groupKey` to `ctx`, so why do I need to reorder them?

[the remark this answers is not in the document]
Comment (zz-jason) [reply expression/aggregation/aggregation.go]: see #4400 for more detail

**Comment by wangyisong1996** [comment]
@zz-jason @hanfei1991 PTAL

File: executor/aggregate.go [hunk 5]
```diff
@@ -44,9 +47,7 @@ type HashAggExec struct {
 func (e *HashAggExec) Close() error {
 	e.groupMap = nil
 	e.groupIterator = nil
-	for _, agg := range e.AggFuncs {
-		agg.Reset()
-	}
+	e.aggCtxsMap = make(aggCtxsMapper, 0)
```
Comment (hanfei1991) [inline executor/aggregate.go]: Setting `e.aggCtxsMap = nil` is better

**Pull request merged**